src/kmeans.jl

```diff
@@ -1,91 +1,125 @@
 # K-means algorithm
 
-#### Interface
-
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
-    assignments::Vector{Int}   # assignments (n)
-    costs::Vector{T}           # costs of the resultant assignments (n)
-    counts::Vector{Int}        # number of samples assigned to each cluster (k)
-    cweights::Vector{Float64}  # cluster weights (k)
-    totalcost::Float64         # total cost (i.e. objective) (k)
-    iterations::Int            # number of elapsed iterations
-    converged::Bool            # whether the procedure converged
+####
+#### Result object
+####
+
+struct KmeansResult{TC<:AbstractFloat,TD<:Real,TCW<:Real} <: ClusteringResult
+    centers::Matrix{TC}       # cluster centers (d x k)
+    assignments::Vector{Int}  # assignments (n)
+    costs::Vector{TD}         # cost of the assignments (n)
+    counts::Vector{Int}       # number of samples assigned to each cluster (k)
+    cweights::Vector{TCW}     # cluster weights (k)
+    totalcost::TD             # total cost (i.e. objective)
+    iterations::Int           # number of elapsed iterations
+    converged::Bool           # whether the procedure converged
 end
 
 const _kmeans_default_init = :kmpp
 const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
-                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
+
+####
+#### Exported kmeans and kmeans! functions
+####
+
+function kmeans!(X::AbstractMatrix{TX}, centers::AbstractMatrix{TC};
+                 weights::Union{Nothing, AbstractVector{TW}}=nothing,
+                 maxiter::Int=_kmeans_default_maxiter,
                  tol::Real=_kmeans_default_tol,
                  display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:Real
+                 distance::SemiMetric=SqEuclidean()
+                 ) where {TX<:Real,TC<:Real,TW<:Real}
+    p, n = size(X)
+    p2, k = size(centers)
 
-    m, n = size(X)
-    m2, k = size(centers)
-    m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
+    p == p2 || throw(DimensionMismatch("Inconsistent array dimensions " *
+                                       "for `X` and `centers`."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
+    if !(weights === nothing)
+        length(weights) == n || throw(DimensionMismatch("Incorrect length " *
+                                                        "of weights."))
+    end
 
     assignments = Vector{Int}(undef, n)
-    costs = Vector{T}(undef, n)
     counts = Vector{Int}(undef, k)
-    cweights = Vector{Float64}(undef, k)
 
-    _kmeans!(X, conv_weights(T, n, weights), centers,
-             assignments, costs, counts, cweights, round(Int, maxiter), tol,
+    TC2 = promote_type(TX, TC)
```

Comment:
I don't think this is the place where `centers` should be converted. It's the user's responsibility to provide it in a correct type since it's an in-place function. Anyway it's quite unlikely that one would be able to use an integer type to store centers. And we already choose an appropriate type in `kmeans` right before calling `kmeans!`, so this seems redundant.

Reply:
No so `kmeans!` is exposed to the user  so can potentially take anything, indeed in the path where it's `kmeans` that's called then `kmeans!` this promotion doesn't happen.

Reply:
I second @nalimilan : it's a very strong Julian convention that `!`-functions modify its corresponding argument in-place (actually, `centers` should be the 1st argument of `kmeans!()`, but I don't know what transition strategy we should use to adopt this API change).
The users are writing code with this convention in mind and violating it will break the their code (e.g. if the users ignore the return value of `kmeans!()` and instead expect the `centers` to be updated).
If it's not possible to modify the argument in-place, the function should fail. In that respect we should restrict `TC` to `TC <: AbstractFloat`: that would also automatically solve `InexactError` issue by throwing much more meaningful `MethodError` beforehand.

Reply:
I'm not convinced that was the point here (the fact that centers gets mutated). 

The point is that the function is exported and therefore a matrix of centers *could* be provided in any format. I don't really see why we would accept anything for `X` but not for `centers` especially when there's virtually no cost incurred by this. 

Note also that `TC <: AbstractFloat` is **not enough** because the centers can be updated by `X_ij * w_j` which could be of a wider type. This is why I suggested this type promotion and relaxed the type to `Real` as the restriction to `AbstractFloat` didn't seem necessary anymore.

Here's a slightly silly example: a user has centers from another code that's somehow in `Float32`, the data `X` and the weights are read as `Float64`, should the updates to centers be downgraded to `Float32` or should the centers be converted to `Float64`?

Reply:
1. Do we agree that `kmeans!()` should only mutate `centers` array provided by the user, unless we want to break the `!`-functions convention and, most likely, the existing code that uses `kmeans!()`?

2. I see these alternatives for what is allowed for centers:
   1) `centers::AbstractMatrix{<:Real}`, calculated as before. If the result of calculation doesn't fit the eltype of `centers` we get `InexactError` exception, which would be confusing for most users
    2) `centers::AbstractMatrix{<:AbstractFloat}`, calculated as before. That will raise `MethodError` for integer `centers`, which is more informative than `InexactError`. (Note that you can assign integer to float, so `centers[i, k] = X[i,j] * w[i,j]` for integer `X` and `w` should be fine). There still could be cases of `InexactError` for certain combinations of `data`, `weights` and `centers` eltypes, but most frequent ones should be handled by the type constraint. If the user needs another eltype for `centers`, he/she should do the conversion after `kmeans!()` call.
    3) `centers::AbstractMatrix{<:Real}`, but we convert the `X[i,j] * w[i,j]` to fit the eltype of `centers` (e.g. `round(Int, ...)`). The drawback is that it's done implicitly, so the users may not realize that these are not the exact centers. Plus this conversion would generate some overhead. @nalimilan Is there any elegant generic way of doing this conversion?

Reply:
Maybe one way to solve both problems would be to deprecate the use of `kmeans!` which doesn't mutate its first argument anyway and maybe suggest a keyword `centers=...` to the `kmeans()` function where that would be taken as the initial centers to start from.

Then there would be no ambiguity with the fact that we convert the `centers` to have the wider type (effectively your iii.) and there would also be no clash with Julia convention with the `!`?